### ISSUE TITLE
javatext: adjust mandatory reqts and focus/reduce the effort

### ISSUE THREAD

**Issue by bjornregnell** [lunduniversity/introprog#423]
@jos-rosenqvist-3812 Thanks for reminding me about this issue!

Javatext currently has these problems:
* students spend too much time on things that don't matter so much
* too many are making everything static

Solution idé:
* update mandatory requirements to make it clear and less cumbersome
* focus the effort on illustrating some important differences between Scala and Java and training the idiomatic java class ''anatomy' with private atributes and constructors and getters setters etc.

Discussion:
* Use this issue thread if discussions on actual solutions are needed.

**Comment by jamil7** [comment]
One of the problems that I remember having 2 years ago when I made my game was that Arrays in Java and Scala didn't play so nicely together. The Scala part of the game was supposed to save the high scores to file and then read it, store it into an Array and let the Java part read it. Well Java didn't like Scala's Arrays and that problem is still unsolved (for my game anyways). 

This can be one of the things to advice students to avoid, or there maybe a very effective solution that I and the assistants missed 2 years ago.

**Comment by jos-rosenqvist-3812** [comment]
I've been sketching on a code skeleton, and I'm wondering if giving the following code to the students would be appropriate (that is, not too easy):

Main.java (already finished – students don't need to implement anything here):
```
import java.util.ArrayList;

public class Main {
    private static ArrayList<Game> playedGames = new ArrayList<>();
    private static boolean quit = false;

    private static void play() {
        System.out.print("Enter your name: ");
        String name = UserInterface.readString();
        Game game = new Game(name);
        game.run();
        playedGames.add(game);
    }

    public static void main(String[] args) {
        while (!quit) {
            System.out.println("Choose an option:");
            int choice = UserInterface.readChoice(new String[] {
                    "1. Play game",
                    "2. View high scores",
                    "3. View high scores for specific player",
                    "4. Quit"
            });

            switch (choice) {
                case 1:
                    play();
                    break;
                case 2:
                    UserInterface.showHighScores(playedGames);
                    break;
                case 3:
                    System.out.print("Enter the player name: ");
                    UserInterface.showHighScores(playedGames, UserInterface.readString());
                    break;
                case 4:
                    quit = true;
                    break;
            }
        }
    }
}
```

Game.java:
```
public class Game {
    public Game(String playerName);

    public String getPlayerName();

    public int getScore();

    public long getElapsedTimeMs();

    public void run();
}
```

UserInterface.scala (maybe not a perfect name, but I couldn't come up with anything better right now...):
```
object UserInterface {
  def readChoice(choices: Array[String]): Int = ???

  def readString(): String = ???

  def showHighScores(scores: java.util.ArrayList[Game]): Unit = ???

  def showHighScores(scores: java.util.ArrayList[Game], player: String): Unit = ???
}
```

That instances of `Game` are kept around after the game finishes essentially forces students to not use static for the important parts of `Game`, and the way the `Game` class is specified gets students to train on the typical Java class structure (albeit without setters). We also get a natural way of passing an `ArrayList` from Java to Scala – a common problem with the current javatext is that it's hard to come up with a way of passing collections between Java and Scala code that actually makes sense and isn't shoehorned in just for the sake of the requirements.

We can keep a lot of the requirements (e.g. what the game code should look like, and that the Scala code should use `java.util.Scanner` for command line input and `scala.collection.JavaConverters` for converting Java collections), but I don't think implementing reading scores from a text file makes a lot of sense with my suggested code skeleton, since you'd need to create `Game` objects (even though you're not going to run them) to store in the `ArrayList`. Maybe it is fine to remove the file IO part of the requirements (people need a lot of time to finish the lab as it currently is, after all), but I would like some feedback on this.

**Comment by bjornregnell** [comment]
@jamil7 there are converters so you can write .asJava se here https://docs.scala-lang.org/overviews/collections/conversions-between-java-and-scala-collections.html
Did that work for you? Perhaps you have a smallest possible example of stuff you had problems with?

**Comment by bjornregnell** [comment]
@jos-rosenqvist-3812 This looks great! It is fine to remove the IO part but it would be nice to still as an extra task to have some presistency on disk, perhaps with introprog.IO, if relevant. Feel free to make a PR on this file [w11-scalajava-lab.tex](https://github.com/lunduniversity/introprog/blob/master/compendium/modules/w11-scalajava-lab.tex) if you have time. 

I will not be able to work much on this before next weeks lectures, but ideally I should have an updated lab printed on paper to hand out at the lecture on 26 of Nov the latest but preferably on the lecture 20th of November...

@jamil7 Perhaps you can offer to review the lab instructions that @jos-rosenqvist-3812 might create if he find the time to pull a lab text into place during this or next week?

**Comment by jos-rosenqvist-3812** [comment]
I have created a pull request: #427

**Comment by bjornregnell** [comment]
@jamil7 There is a pull request now here: #427. If you are interested, you are also welcome to give feedback on the PR in that conversation thread.

**Comment by jamil7** [comment]
@bjornregnell I jus took a look at the code and it seems that most of the structural problems are solved, especially with the arrays. I will try to migrate my old game to fit this new structure and see if my old problems will resolve themselves :D 

The issue that I faced was using Scala data structures in Java. asJava/asScala didn't really work on those. I will try to replicate the problem and show you a snippet of it. But, thinking about it now it could be circumvented by sticking to one set of data structures (either Java's or Scala's) and not using them interchangeably. Jos' code does exactly this by using Java's arrays in Scala classes.

**Comment by jamil7** [comment]
@bjornregnell Okay, so the migration was successful but a bit hacky. When I did the lab back in the day there weren't any skeleton classes so I had to adjust the program to fit Jos' skeleton. 

From my own experience, I found that the skeleton code and the re-write of the lab fixed a lot of the problems that existed before, especially static/non-static classes and such. 

The problem that I had with arrays seems to have solved itself in som magical way, or rather that I actually know how Scala works now compared to the time that I did the lab 2 years ago where the whole language felt like fp mumbo-jumbo and syntactical sugar and magic.

However, if I do remember correctly one of the main points of this lab was to save the data to the file system and retrieve it to display the high scores and such. I've added this functionality, but I'm not sure if it is a requirement any longer. I would personally encourage making it a requirement just to give the students a taste of how saving files can be done. Eventually giving the choice between going through the high scores array and saving the scores as text, serialising the array and writing it to a file as a Java object, or even introducing serialisation to JSON as an extra challenge, considering how popular JSON is these days. 

Another point which the lab didn't explicitly touch on, is the generics in Java, or in general. Back in my day we were encouraged to use generics in this lab, but it was mostly related to reading/writing the high scores to the file system. From my understanding, generic functions and types where introduces in chapter 7 and 8 of the book, so it would by nice to use them a bit more and more importantly make the students write their own generic methods and not just use some that are pre-made.

To summarise the skeleton is beautifully simple enough for anyone to get started with, while being instructive enough to guide away from common mistakes and bad-practice. Amazing job @jos-rosenqvist-3812 ! However, the inclusion of saving the high scores to the file system would've made a great addition, and adding a sprinkle of generics would've made the lab even sweeter.

**Comment by jamil7** [comment]
@bjornregnell After heavily refactoring my own code I ended up with a neat little IO class, which can be easily utilised to save and load the high scores as Java objects. It can be easily integrated with the lab code as a Java class as in the following example: 

`class IOHandler {
     static void objectWriter(File file, List<Game> list) {
        try {
            // TODO: implement an object writer
            // Tip: look up ObjectOutputStream
        } catch () {
          // TODO: catch some errors!
        }
    }

     static List<Game> objectReader(File file) {
        try {
            // TODO: implement an object writer
            // Tip: look up ObjectInputStream this time
        }  catch () {
            TODO: catch some errors!
        }
    }
}`

The class can be re-written to a stand alone Scala class to use `asScala` and `asJava` a bit more, and since it's a relatively small class it can be even merged with `UserInterface.scala`. 

`ObjectWriter() `can be made more generic as well, accepting an `Object` as a second parameter instead for `List<Game>`.
